integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java

```diff
@@ -0,0 +1,250 @@
+        assertEquals(ENFORCING, p.enforcementMode());
+        assertTrue(p.claimInformationPoint().complexConfig().isEmpty());
+        assertFalse(p.claimInformationPoint().simpleConfig().isEmpty());
+        assertEquals(1, p.claimInformationPoint().simpleConfig().size());
+        assertFalse(p.claimInformationPoint().simpleConfig().get("key").isEmpty());
+        assertEquals("sub-value", p.claimInformationPoint().simpleConfig().get("key").get("sub-key"));
+        assertFalse(p.name().isEmpty());
+        assertEquals("New Permission", p.name().get());
+        assertEquals(1, p.paths().get().size());
+        assertEquals("/api/one", p.paths().get().get(0));
+        assertEquals(1, p.methods().size());
+        method = p.methods().get("trace");
+        assertNull(method);
+        method = p.methods().get("get");
+        // we didn't add scopes, and we are yet to require them
+        assertTrue(method.scopes().contains("scope5"));
+        assertEquals("GET", method.method());
+        // default scope enforcement is ALL
+        assertEquals(PolicyEnforcerConfig.ScopeEnforcementMode.DISABLED, method.scopesEnforcementMode());
+        // in addition to the GET and PUT and POST methods
+        builder = KeycloakPolicyEnforcerTenantConfig.builder(config);
+        path = builder.setPaths("/api/one");
+        path.setPost("1", "2");
+        path.setPut(PolicyEnforcerConfig.ScopeEnforcementMode.ANY, "3", "4");
+        config = builder.build();
+        var pathConfig = config.policyEnforcer().paths().get("/api/one");
+        assertEquals(3, pathConfig.methods().size());
+        var putMethod = pathConfig.methods().get("put");
+        assertNotNull(putMethod);
+        assertTrue(putMethod.scopes().contains("3"));
+        assertTrue(putMethod.scopes().contains("4"));
+        assertEquals("PUT", putMethod.method());
+        assertEquals(PolicyEnforcerConfig.ScopeEnforcementMode.ANY, putMethod.scopesEnforcementMode());
+        var postMethod = pathConfig.methods().get("post");
+        assertNotNull(postMethod);
+        assertEquals(PolicyEnforcerConfig.ScopeEnforcementMode.ALL, postMethod.scopesEnforcementMode());
+        assertEquals("POST", postMethod.method());
+        assertTrue(postMethod.scopes().contains("1"));
+        assertTrue(postMethod.scopes().contains("2"));
+        // test config with multiple path patterns
+        config = KeycloakPolicyEnforcerTenantConfig.builder().setPaths("/x", "/y").setEnforcementMode(PERMISSIVE).build();
+        pathConfig = config.policyEnforcer().paths().entrySet().stream().findAny().get().getValue();
+        assertEquals(2, pathConfig.paths().get().size());
+        assertTrue(pathConfig.paths().get().contains("/x"));
+        assertTrue(pathConfig.paths().get().contains("/y"));
+        assertEquals(PERMISSIVE, pathConfig.enforcementMode());
+    }
+
+    private static void assertEveryConfigPropertyCanBeSet() {
+        var builder = KeycloakPolicyEnforcerTenantConfig.builder()
```

Comment:
@michalvavrik It looks fine, but indeed, my earlier suggestion to use `set` does not look great, instead,
```
 KeycloakPolicyEnforcerTenantConfig.builder()
                .enforcementMode(DISABLED)
                .claimInformationPoint(Map.of("one", Map.of("two", "three")), Map.of("four", Map.of()))
                .connectionPoolSize(-1)
                .lazyLoadPaths(false)
                .pathCache(5, 2)
                .httpMethodAsScope(true)
                .build();
```
would be nicer, sorry for suggesting adding `set`.

Can you also please give me a favor and add one more test, showing the same properties but which will expand shortcuts for the groups like `.pathCache(5, 2)` into `.pathCache().prop1(5).prop2(2)`.

I'm sorry it is a bit time consuming, but IMHO the more compact the sequence the better the user experience will be, as we agreed, they would provide a template for the OidcTenantConfig and OidcClientConfig. Thanks for your patience :-)

Reply:
np, on it

> Can you also please give me a favor and add one more test, showing the same properties but which will expand shortcuts for the groups like .pathCache(5, 2) into .pathCache().prop1(5).prop2(2).

well, that is no longer possible. I leveraged the fact that maxEntries and lifespan has different data type so you can do either `pathCache(5,2)` or `pathCache(5)` or `pathCache(2)`. You can read method parameter names and I can add a javadoc (though TBH I think the generated javadoc by IDE is more then sufficient). IMO it is not confusing because whatever you do, you need to read method parameters in order to know what you are setting. Personally I don't know almost any methods by heart.

Reply:
![image](https://github.com/quarkusio/quarkus/assets/43821672/2e615de1-5610-4471-b4ea-0eebab46b51f)

Reply:
![image](https://github.com/quarkusio/quarkus/assets/43821672/2e615de1-5610-4471-b4ea-0eebab46b51f)

Reply:
I've added `DynamicTenantConfigPolicyEnforcerTest#assertPathCacheConfigOnly` that shows all the path cache combinations.

Reply:
> It looks fine, but indeed, my earlier suggestion to use set does not look great, instead,

I've dropped `set` and added path cache tests that address questions previous. I'll close this comment thread as i believe it will be easier to discuss possible new questions separately.

Reply:
@michalvavrik It looks nice, but losing `.pathCache().a().b()` and the same for other groups is not great IMHO.

Do you mean that with `.pathCache().a().b()` we have to close it with `build()` ?

So either

`KeycloakPolicyBuilder.builder()
  .pathCache(1, 2)
  .build()`

or 
```
KeycloakPolicyBuilder.builder()
  .pathCache().a(1).b(2).build()
  .build()
```
?

See if the group has say 7 or more properties, a few of such groups are available in OIDC, then users would have to type `builder.myGroup(1, null, 3, null) as it is nearly impossible to offer all possible overrides for users to choose the most suitable one where they can avoid setting `null`.

So not sure we can really avoid supporting itin addition to the shortcuts like `.pathCache(1, 2)`

May be a tiny optimization is possible in this case:

```
KeycloakPolicyBuilder
  .pathCache().a(1).b(2).build()
  .someOtherGroup().c(1).d(2).e().build()
  .build()
```

where `KeycloakPolicyBuilder.pathCache()` inits the main builder and starts the specific group builder so the users do not have to type  `.builder()`.

Reply:
> Do you mean that with .pathCache().a().b() we have to close it with build() ?

there is no such thing like `pathCache().a().b()`, only configmapping has it, we have 3 setters: `pathCache(1,2)`, `pathCache(1L)`, `pathCache(2)`.

> See if the group has say 7 or more properties, a few of such groups are available in OIDC, then users would have to type builder.myGroup(1, null, 3, null) as it is nearly impossible to offer all possible overrides for users to choose the most suitable one where they can avoid setting null.

currently there is only one group like that, it is `Path` and for that, I created it's own builder because I don't like `builder.myGroup(1, null, 3, null)`. so it works like this:

```
var pathBuilder = builder.namedPaths("p1", "path");
pathBuilder.enforcementMode(PolicyEnforcerConfig.EnforcementMode.PERMISSIVE);
pathBuilder.permissionName("n1");
pathBuilder.method("method1", PolicyEnforcerConfig.ScopeEnforcementMode.DISABLED, "scope1", "scopes2");
var config = builder.build();
```

What I like about this is that there is no `build`, there is no `saving` or `end()`, it is build as soon as you build the parent builder and then, it builds immutable configuration interface. You can keep using the original builder without any risk.

> So not sure we can really avoid supporting itin addition to the shortcuts like .pathCache(1, 2)

As long as we can settle on something, it's absolutely fine to create these builders again for users...

> May be a tiny optimization is possible in this case:

That's not a tiny optimization, it will require me to create all of these sub-builders again. I thought you wanted to drop them. Your example has 2 `build()` again, which I though you wrote somewhere is not necessary. I am happy to do it if you are certain @sberyozkin .

I don't see appeal in doing `pathCache().a(1).b(2).build().some...` over `pathCache(1, 2)` but it is important to get it right, so if you are absolutely certain @sberyozkin , let's do it.

Reply:
Hi @michalvavrik 

I did not mean `pathCache().a(1).b(2).build()` was better than `pathCache(1, 2)` but the former is easier to understand by just looking at the existing code, what `a` and `b` means, `pathCache(1, 2)` is clear enough when we type it in IDE, but not afterwards.

> currently there is only one group like that, it is Path and for that, I created it's own builder because I don't like builder.myGroup(1, null, 3, null). so it works like this:
```
 var pathBuilder = builder.namedPaths("p1", "path");
pathBuilder.enforcementMode(PolicyEnforcerConfig.EnforcementMode.PERMISSIVE);
pathBuilder.permissionName("n1");
pathBuilder.method("method1", PolicyEnforcerConfig.ScopeEnforcementMode.DISABLED, "scope1", "scopes2");
var config = builder.build();
```

Sure, so here you close everything with the parent's builder's `build()`, but users can't do it all in a single sequence so a builder specific `build()` would only be needed for the single sequence expression. 

IMHO, it should be possible to build every group in one of several ways and it should be consistent for every group:
* builder.group1(1, 2,3,4,5).group2(5,6,7,8,9).build(); // and add as many reasonable overloads for a given group as needed
* builder.group1().a1(1).b1(1).end().group2().a2(2).b2(2).end().build(); 

I'm using `end()` following your comment above to indicate a group specific end - users won't have to use it - it is only there to support a single sequence. But as you showed, users would then also be able to skip them with a single `build()` if they want to skip group specific builds. 

>That's not a tiny optimization, it will require me to create all of these sub-builders again. 

There I was only referring to the option where you skip initial `.builder()` - but if you agree to the earlier suggestions, you would be creating individual group builders in any case...

I agree it all may look  like a bit of over-engineering, but IMHO it is worth trying to make it as optimal and consistent as we can, like we agreed, it is a template for providing future OIDC builders

Reply:
ok, I'll do it.